File: cpp/tests/transport/sipo_test_linux/sipo_test_linux.cpp
#include <cstdio>
#include <cstdlib>
#include <gpiod.h>
#include "SiPoTransportLinux.h"

#ifndef TEST_GPIO_CHIP
#define TEST_GPIO_CHIP    "/dev/gpiochip0"
#endif
#ifndef TEST_SER_IN_LINE
#define TEST_SER_IN_LINE  19
#endif
#ifndef TEST_SRCK_LINE
#define TEST_SRCK_LINE    26
#endif
#ifndef TEST_RCK_LINE
#define TEST_RCK_LINE     5
#endif
#ifndef TEST_SRCLR_LINE
#define TEST_SRCLR_LINE   6
#endif
#ifndef TEST_G_LINE
#define TEST_G_LINE       13
#endif

static int passed = 0;
static int failed = 0;

static void check_true(const char* label, bool condition) {
    if (condition) { printf("PASS %s\n", label); passed++; }
    else           { printf("FAIL %s\n", label); failed++; }
}

int main() {
    struct gpiod_chip* chip = gpiod_chip_open(TEST_GPIO_CHIP);
    if (!chip) { perror("gpiod_chip_open"); return 2; }

    struct gpiod_line* ser_in = gpiod_chip_get_line(chip, TEST_SER_IN_LINE);
    struct gpiod_line* srck   = gpiod_chip_get_line(chip, TEST_SRCK_LINE);
    struct gpiod_line* rck    = gpiod_chip_get_line(chip, TEST_RCK_LINE);
    struct gpiod_line* srclr  = gpiod_chip_get_line(chip, TEST_SRCLR_LINE);
    struct gpiod_line* g      = gpiod_chip_get_line(chip, TEST_G_LINE);

    gpiod_line_request_output(ser_in, "sipo_test", 0);
    gpiod_line_request_output(srck,   "sipo_test", 0);
    gpiod_line_request_output(rck,    "sipo_test", 0);
    gpiod_line_request_output(srclr,  "sipo_test", 1);
    gpiod_line_request_output(g,      "sipo_test", 0);

    SiPoTransportLinux transport(ser_in, srck, rck, srclr, g);

    uint8_t data1[] = { 0xA5 };
    transport.write(data1, sizeof(data1));
    check_true("write accepted", true);

    uint8_t data2[] = { 0x00, 0xFF };
    transport.write(data2, sizeof(data2));
    check_true("write multi-byte accepted", true);

    bool threw = false;
    try {
        transport.clear();
    } catch (...) {
        threw = true;
    }
    check_true("clear accepted", !threw);

    try {
        transport.set_output_enable(false);
        transport.set_output_enable(true);
    } catch (...) {
        threw = true;
    }
    check_true("set_output_enable accepted", !threw);

    transport.close();
    check_true("close accepted", true);

    gpiod_chip_close(chip);

    printf("===DONE: %d passed, %d failed===\n", passed, failed);
    return failed == 0 ? 0 : 1;
}

File: python/tests/transport/sipo_test_linux.py
import os
import sys

import gpiod
from periph.transport.sipo_linux import SiPoTransport

CHIP = os.environ.get('GPIO_CHIP', '/dev/gpiochip0')
MODE = os.environ.get('SIPO_MODE', 'sw')  # 'sw' (bit-bang) or 'hw' (spidev)

RCK    = int(os.environ.get('SIPO_RCK',    '5'))
SRCLR  = int(os.environ.get('SIPO_SRCLR',  '6'))
G      = int(os.environ.get('SIPO_G',      '13'))
SER_IN = int(os.environ.get('SIPO_SER_IN', '19'))
SRCK   = int(os.environ.get('SIPO_SRCK',   '26'))

SPI_BUS    = int(os.environ.get('SIPO_SPI_BUS', '0'))
SPI_DEVICE = int(os.environ.get('SIPO_SPI_DEVICE', '0'))

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


offsets = [RCK, SRCLR, G]
if MODE == 'sw':
    offsets += [SER_IN, SRCK]

chip = gpiod.Chip(CHIP)
settings = gpiod.LineSettings(direction=gpiod.line.Direction.OUTPUT)
request = chip.request_lines(consumer='sipo_test', config={o: settings for o in offsets})

if MODE == 'hw':
    transport = SiPoTransport(request, RCK, bus_num=SPI_BUS, device_num=SPI_DEVICE,
                               srclr_offset=SRCLR, g_offset=G)
else:
    transport = SiPoTransport(request, RCK, ser_in_offset=SER_IN, srck_offset=SRCK,
                               srclr_offset=SRCLR, g_offset=G)

transport.write(bytes([0xA5]))
check_true('write accepted', True)

transport.write(bytes([0x00, 0xFF]))
check_true('write multi-byte accepted', True)

transport.clear()
check_true('clear accepted', True)

transport.set_output_enable(False)
check_true('set_output_enable(False) accepted', True)

transport.set_output_enable(True)
check_true('set_output_enable(True) accepted', True)

transport.close()
check_true('close accepted', True)

print('===DONE: {} passed, {} failed==='.format(passed, failed))
sys.exit(0 if failed == 0 else 1)
